File: ios/Classes/Data/DataSource/FileStorageDataSource.swift
import Foundation

/// Data source for working with file storage
class FileStorageDataSource {
    private let storageDir: URL
    
    init() {
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        storageDir = documentsURL.appendingPathComponent("background_http_client", isDirectory: true)
        
        // Create directories if they do not exist
        try? FileManager.default.createDirectory(at: storageDir, withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: requestsDir, withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: responsesDir, withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: statusDir, withIntermediateDirectories: true)
        try? FileManager.default.createDirectory(at: bodiesDir, withIntermediateDirectories: true)
    }
    
    private var requestsDir: URL {
        storageDir.appendingPathComponent("requests", isDirectory: true)
    }
    
    private var responsesDir: URL {
        storageDir.appendingPathComponent("responses", isDirectory: true)
    }
    
    private var statusDir: URL {
        storageDir.appendingPathComponent("status", isDirectory: true)
    }
    
    private var bodiesDir: URL {
        storageDir.appendingPathComponent("request_bodies", isDirectory: true)
    }
    
    /// Saves a request to a file and returns task information
    func saveRequest(request: HttpRequest, requestId: String, registrationDate: Int64) throws -> TaskInfo {
        // Save body to a separate file if present
        if let body = request.body, let bodyData = body.data(using: .utf8) {
            let bodyFile = bodiesDir.appendingPathComponent("\(requestId).body")
            try bodyData.write(to: bodyFile)
        }
        
        // Save request as JSON
        let requestFile = requestsDir.appendingPathComponent("\(requestId).json")
        let requestData: [String: Any] = [
            "url": request.url,
            "method": request.method,
            "headers": request.headers ?? [:],
            "body": request.body ?? "",
            "queryParameters": request.queryParameters ?? [:],
            "timeout": request.timeout ?? 120,
            "multipartFields": request.multipartFields ?? [:],
            "multipartFiles": request.multipartFiles?.mapValues { file in
                [
                    "filePath": file.filePath,
                    "filename": file.filename ?? "",
                    "contentType": file.contentType ?? ""
                ]
            } ?? [:],
            "requestId": requestId,
            "retries": request.retries ?? 0,
            "stuckTimeoutBuffer": request.stuckTimeoutBuffer ?? 60,
            "queueTimeout": request.queueTimeout ?? 600
        ]
        
        let jsonData = try JSONSerialization.data(withJSONObject: requestData)
        try jsonData.write(to: requestFile)
        
        // Save initial status
        try saveStatus(requestId: requestId, status: .inProgress, startTime: registrationDate)
        
        return TaskInfo(
            id: requestId,
            status: .inProgress,
            path: requestFile.path,
            registrationDate: registrationDate,
            responseJson: nil
        )
    }
    
    /// Loads task information
    func loadTaskInfo(requestId: String) -> TaskInfo? {
        let requestFile = requestsDir.appendingPathComponent("\(requestId).json")
        guard FileManager.default.fileExists(atPath: requestFile.path) else {
            return nil
        }
        
        let status = loadStatus(requestId: requestId) ?? .inProgress
        // Get registration date from status file if present, otherwise use lastModified
        let registrationDate = loadRegistrationDate(requestId: requestId) ?? {
            if let attributes = try? FileManager.default.attributesOfItem(atPath: requestFile.path),
               let modificationDate = attributes[.modificationDate] as? Date {
                return Int64(modificationDate.timeIntervalSince1970 * 1000)
            } else {
                return Int64(Date().timeIntervalSince1970 * 1000)
            }
        }()
        
        return TaskInfo(
            id: requestId,
            status: status,
            path: requestFile.path,
            registrationDate: registrationDate,
            responseJson: nil
        )
    }
    
    /// Loads registration date from the status file
    private func loadRegistrationDate(requestId: String) -> Int64? {
        let statusFile = statusDir.appendingPathComponent("\(requestId).json")
        guard FileManager.default.fileExists(atPath: statusFile.path),
              let data = try? Data(contentsOf: statusFile),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let startTime = json["startTime"] as? Int64 else {
            return nil
        }
        return startTime
    }
    
    /// Loads task response
    func loadTaskResponse(requestId: String) -> TaskInfo? {
        guard var taskInfo = loadTaskInfo(requestId: requestId) else {
            return nil
        }
        
        let responseFile = responsesDir.appendingPathComponent("\(requestId).json")
        guard FileManager.default.fileExists(atPath: responseFile.path) else {
            return taskInfo
        }
        
        if let data = try? Data(contentsOf: responseFile),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            return TaskInfo(
                id: taskInfo.id,
                status: taskInfo.status,
                path: taskInfo.path,
                registrationDate: taskInfo.registrationDate,
                responseJson: json
            )
        }
        
        return taskInfo
    }
    
    /// Saves task status
    func saveStatus(requestId: String, status: RequestStatus, startTime: Int64? = nil) throws {
        let statusFile = statusDir.appendingPathComponent("\(requestId).json")
        var statusData: [String: Any] = [
            "requestId": requestId,
            "status": status.rawValue
        ]
        
        if let startTime = startTime {
            statusData["startTime"] = startTime
        }
        
        let jsonData = try JSONSerialization.data(withJSONObject: statusData)
        try jsonData.write(to: statusFile)
    }
    
    /// Loads task status
    func loadStatus(requestId: String) -> RequestStatus? {
        let statusFile = statusDir.appendingPathComponent("\(requestId).json")
        guard FileManager.default.fileExists(atPath: statusFile.path),
              let data = try? Data(contentsOf: statusFile),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let statusValue = json["status"] as? Int,
              let status = RequestStatus(rawValue: statusValue) else {
            return nil
        }
        
        return status
    }
    
    /// Saves server response
    func saveResponse(
        requestId: String,
        statusCode: Int,
        headers: [String: String],
        body: String?,
        responseFilePath: String?,
        status: RequestStatus,
        error: String?
    ) throws {
        let responseFile = responsesDir.appendingPathComponent("\(requestId).json")
        let responseData: [String: Any] = [
            "requestId": requestId,
            "statusCode": statusCode,
            "headers": headers,
            "body": body ?? "",
            "responseFilePath": responseFilePath ?? "",
            "status": status.rawValue,
            "error": error ?? ""
        ]
        
        let jsonData = try JSONSerialization.data(withJSONObject: responseData)
        try jsonData.write(to: responseFile)
        
        // Update status
        try saveStatus(requestId: requestId, status: status)
    }
    
    /// Deletes all files associated with a task
    func deleteTaskFiles(requestId: String) -> Bool {
        var deleted = true
        
        // Delete request file
        let requestFile = requestsDir.appendingPathComponent("\(requestId).json")
        if FileManager.default.fileExists(atPath: requestFile.path) {
            try? FileManager.default.removeItem(at: requestFile)
        } else {
            deleted = false
        }
        
        // Delete body file
        let bodyFile = bodiesDir.appendingPathComponent("\(requestId).body")
        try? FileManager.default.removeItem(at: bodyFile)
        
        // Delete JSON response file
        let responseJsonFile = responsesDir.appendingPathComponent("\(requestId).json")
        try? FileManager.default.removeItem(at: responseJsonFile)
        
        // Delete response data file
        let responseDataFile = responsesDir.appendingPathComponent("\(requestId)_response.txt")
        try? FileManager.default.removeItem(at: responseDataFile)
        
        // Delete status file
        let statusFile = statusDir.appendingPathComponent("\(requestId).json")
        try? FileManager.default.removeItem(at: statusFile)
        
        return deleted
    }
    
    /// Checks if a task exists
    func taskExists(requestId: String) -> Bool {
        let requestFile = requestsDir.appendingPathComponent("\(requestId).json")
        return FileManager.default.fileExists(atPath: requestFile.path)
    }
    
    /// Gets a list of all task IDs from the file system
    func getAllTaskIds() -> [String] {
        guard let files = try? FileManager.default.contentsOfDirectory(at: requestsDir, includingPropertiesForKeys: nil) else {
            return []
        }
        return files
            .filter { $0.pathExtension == "json" }
            .map { $0.deletingPathExtension().lastPathComponent }
    }
    
    /// Loads request data from file
    func loadRequestData(requestId: String) -> RequestData? {
        let requestFile = requestsDir.appendingPathComponent("\(requestId).json")
        guard FileManager.default.fileExists(atPath: requestFile.path),
              let data = try? Data(contentsOf: requestFile),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        
        // Parse multipartFiles
        var multipartFiles: [String: MultipartFile]?
        if let filesDict = json["multipartFiles"] as? [String: [String: Any]] {
            multipartFiles = [:]
            for (key, value) in filesDict {
                if let filePath = value["filePath"] as? String {
                    multipartFiles?[key] = MultipartFile(
                        filePath: filePath,
                        filename: value["filename"] as? String,
                        contentType: value["contentType"] as? String
                    )
                }
            }
        }
        
        return RequestData(
            url: json["url"] as? String ?? "",
            method: json["method"] as? String ?? "GET",
            headers: json["headers"] as? [String: String],
            body: json["body"] as? String,
            queryParameters: json["queryParameters"] as? [String: String],
            timeout: json["timeout"] as? Int,
            multipartFields: json["multipartFields"] as? [String: String],
            multipartFiles: multipartFiles,
            retries: json["retries"] as? Int
        )
    }
}

/// Structure for storing request data
struct RequestData {
    let url: String
    let method: String
    let headers: [String: String]?
    let body: String?
    let queryParameters: [String: String]?
    let timeout: Int?
    let multipartFields: [String: String]?
    let multipartFiles: [String: MultipartFile]?
    let retries: Int?
}

File: android/src/main/kotlin/com/tsitser/background_http_plugin/data/datasource/FileStorageDataSource.kt
package com.tsitser.background_http_plugin.data.datasource

import android.content.Context
import com.tsitser.background_http_plugin.domain.entity.HttpRequest
import com.tsitser.background_http_plugin.domain.entity.RequestStatus
import com.tsitser.background_http_plugin.domain.entity.TaskInfo
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Data source for working with file storage.
 */
class FileStorageDataSource(private val context: Context) {

    private val storageDir: File by lazy {
        File(context.filesDir, "background_http_client").apply {
            if (!exists()) mkdirs()
        }
    }

    private val requestsDir: File by lazy {
        File(storageDir, "requests").apply {
            if (!exists()) mkdirs()
        }
    }

    private val responsesDir: File by lazy {
        File(storageDir, "responses").apply {
            if (!exists()) mkdirs()
        }
    }

    private val statusDir: File by lazy {
        File(storageDir, "status").apply {
            if (!exists()) mkdirs()
        }
    }

    private val bodiesDir: File by lazy {
        File(storageDir, "request_bodies").apply {
            if (!exists()) mkdirs()
        }
    }

    /**
     * Saves a request to a file and returns task information.
     */
    fun saveRequest(request: HttpRequest, requestId: String, registrationDate: Long): TaskInfo {
        // Save body to a separate file if present
        if (request.body != null) {
            val bodyFile = File(bodiesDir, "$requestId.body")
            bodyFile.writeText(request.body)
        }

        // Save request as JSON
        val requestFile = File(requestsDir, "$requestId.json")
        val requestJson = JSONObject().apply {
            put("url", request.url)
            put("method", request.method)
            put("headers", request.headers?.let { JSONObject(it as Map<*, *>) } ?: JSONObject())
            put("body", request.body ?: "")
            put("queryParameters", request.queryParameters?.let { JSONObject(it as Map<*, *>) } ?: JSONObject())
            put("timeout", request.timeout ?: 120)
            put("multipartFields", request.multipartFields?.let { JSONObject(it as Map<*, *>) } ?: JSONObject())
            put("multipartFiles", request.multipartFiles?.let { files ->
                JSONObject().apply {
                    files.forEach { (key, file) ->
                        put(key, JSONObject().apply {
                            put("filePath", file.filePath)
                            put("filename", file.filename ?: "")
                            put("contentType", file.contentType ?: "")
                        })
                    }
                }
            } ?: JSONObject())
            put("requestId", requestId)
            put("retries", request.retries ?: 0)
            put("stuckTimeoutBuffer", request.stuckTimeoutBuffer ?: 60)
            put("queueTimeout", request.queueTimeout ?: 600)
        }
        requestFile.writeText(requestJson.toString())

        // Save initial status
        saveStatus(requestId, RequestStatus.IN_PROGRESS, registrationDate)

        return TaskInfo(
            id = requestId,
            status = RequestStatus.IN_PROGRESS,
            path = requestFile.absolutePath,
            registrationDate = registrationDate
        )
    }

    /**
     * Loads task information.
     */
    fun loadTaskInfo(requestId: String): TaskInfo? {
        val requestFile = File(requestsDir, "$requestId.json")
        if (!requestFile.exists()) {
            return null
        }

        val status = loadStatus(requestId) ?: RequestStatus.IN_PROGRESS
        // Get registration date from status file if present, otherwise use lastModified
        val registrationDate = loadRegistrationDate(requestId) ?: requestFile.lastModified()

        return TaskInfo(
            id = requestId,
            status = status,
            path = requestFile.absolutePath,
            registrationDate = registrationDate
        )
    }

    /**
     * Loads registration date from the status file.
     */
    private fun loadRegistrationDate(requestId: String): Long? {
        val statusFile = File(statusDir, "$requestId.json")
        if (!statusFile.exists()) {
            return null
        }

        return try {
            val jsonString = statusFile.readText()
            val json = JSONObject(jsonString)
            if (json.has("startTime")) {
                json.getLong("startTime")
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Loads task response.
     */
    fun loadTaskResponse(requestId: String): TaskInfo? {
        val taskInfo = loadTaskInfo(requestId) ?: return null

        val responseFile = File(responsesDir, "$requestId.json")
        if (!responseFile.exists()) {
            return taskInfo
        }

        val responseJson = try {
            val jsonString = responseFile.readText()
            val json = JSONObject(jsonString)
            json.toMap()
        } catch (e: Exception) {
            null
        }

        return taskInfo.copy(responseJson = responseJson)
    }

    /**
     * Saves task status.
     */
    fun saveStatus(requestId: String, status: RequestStatus, startTime: Long? = null) {
        val statusFile = File(statusDir, "$requestId.json")
        val statusJson = JSONObject().apply {
            put("requestId", requestId)
            put("status", status.value)
            if (startTime != null) {
                put("startTime", startTime)
            }
        }
        statusFile.writeText(statusJson.toString())
    }

    /**
     * Loads task status.
     */
    fun loadStatus(requestId: String): RequestStatus? {
        val statusFile = File(statusDir, "$requestId.json")
        if (!statusFile.exists()) {
            return null
        }

        return try {
            val jsonString = statusFile.readText()
            val json = JSONObject(jsonString)
            val statusValue = json.getInt("status")
            enumValues<RequestStatus>().firstOrNull { it.value == statusValue }
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Saves server response.
     */
    fun saveResponse(
        requestId: String,
        statusCode: Int,
        headers: Map<String, String>,
        body: String?,
        responseFilePath: String?,
        status: RequestStatus,
        error: String?
    ) {
        val responseFile = File(responsesDir, "$requestId.json")
        val responseJson = JSONObject().apply {
            put("requestId", requestId)
            put("statusCode", statusCode)
            put("headers", JSONObject(headers as Map<*, *>))
            put("body", body ?: "")
            put("responseFilePath", responseFilePath ?: "")
            put("status", status.value)
            put("error", error ?: "")
        }
        responseFile.writeText(responseJson.toString())

        // Update status
        saveStatus(requestId, status)
    }

    /**
     * Deletes all files associated with a task.
     */
    fun deleteTaskFiles(requestId: String): Boolean {
        var deleted = true

        // Delete request file
        File(requestsDir, "$requestId.json").takeIf { it.exists() }?.delete() ?: run { deleted = false }

        // Delete body file
        File(bodiesDir, "$requestId.body").takeIf { it.exists() }?.delete()

        // Delete JSON response file
        File(responsesDir, "$requestId.json").takeIf { it.exists() }?.delete()

        // Delete response data file
        File(responsesDir, "${requestId}_response.txt").takeIf { it.exists() }?.delete()

        // Delete status file
        File(statusDir, "$requestId.json").takeIf { it.exists() }?.delete()

        return deleted
    }

    /**
     * Checks whether a task exists.
     */
    fun taskExists(requestId: String): Boolean {
        return File(requestsDir, "$requestId.json").exists()
    }

    /**
     * Gets a list of all task IDs from the file system.
     */
    fun getAllTaskIds(): List<String> {
        return requestsDir.listFiles()
            ?.filter { it.isFile && it.name.endsWith(".json") }
            ?.map { it.name.removeSuffix(".json") }
            ?: emptyList()
    }

    /**
     * Helper function for converting JSONObject to Map.
     */
    private fun JSONObject.toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        val keys = this.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            val value = this.get(key)
            map[key] = when (value) {
                is JSONObject -> value.toMap()
                is JSONArray -> value.toList()
                else -> value
            }
        }
        return map
    }

    private fun JSONArray.toList(): List<Any> {
        val list = mutableListOf<Any>()
        for (i in 0 until this.length()) {
            val value = this.get(i)
            list.add(
                when (value) {
                    is JSONObject -> value.toMap()
                    is JSONArray -> value.toList()
                    else -> value
                }
            )
        }
        return list
    }
}

